lib/GitterUser.js

```diff
@@ -0,0 +1,128 @@
+"use strict";
+
+var Promise = require("bluebird");
+
+var rp = require("request-promise");
+
+// miliseconds of interval between presence=online updates
+var ONLINE_PERIOD = 4 * 60 * 1000;
+
+// miliseconds of grace period before users count as offline
+var OFFLINE_GRACE = 3 * 60 * 1000;
+
+function GitterUser(bridge, remoteUserModel) {
+  this._bridge = bridge;
+  this._remoteUserModel = remoteUserModel;
+
+  this._presenceRooms = {};
+  this._ispresent = false;
+  this._presenceinterval = null;
+  this._presencedelay = null;
+}
+
+GitterUser.prototype.gitterId = function() {
+  return this._remoteUserModel.getId();
+};
+
+GitterUser.prototype.gitterUsername = function() {
+  return this._remoteUserModel.get("username");
+};
+
+GitterUser.prototype.getIntent = function() {
+  return this._intent = this._intent || (
+      this._bridge.getIntentFromLocalpart('gitter_' + this.gitterUsername()));
+};
+
+GitterUser.prototype.update = function(user) {
+  return Promise.all([
+      this.updateDisplayName(user.displayName),
+      this.updateAvatar(user.avatarUrlMedium)
+  ]);
+};
+
+GitterUser.prototype.updateDisplayName = function(name) {
+  if (this._remoteUserModel.get("display_name") == name)
+    return Promise.resolve();
+
+  var intent = this.getIntent();
+
+  return intent.setDisplayName(name + ' (Gitter)').then(() => {
+    this._remoteUserModel.set("display_name", name);
+
+    return this._bridge.putRemoteUser(this._remoteUserModel);
+  });
+};
+
+GitterUser.prototype.updateAvatar = function(url) {
+  if (this._remoteUserModel.get("avatar_url") == url)
+    return Promise.resolve();
+
+  console.log("Updating " + this.gitterUsername() + " avatar image from " + url);
+
+  var intent = this.getIntent();
+
+  return rp({
+    uri: url,
+    resolveWithFullResponse: true,
+    encoding: null
+  }).then((response) => {
+    var content_type = response.headers["content-type"];
+
+    return intent.getClient().uploadContent({
+      stream: new Buffer(response.body, "binary"),
+      name: "avatar.jpg",
+      type: content_type,
```

Comment:
Ah. :) That's me being a little subtle.
The Matrix content API requires there to be a name but it really doesn't care what that name is. I just threw that in to keep it happy - I'm not sure the type really matters, the type could be anything.